File: cpp/120_Triangle.py
class Solution:
    def minimumTotal(self, tri: List[List[int]]) -> int:
        m = len(tri)

        dp = tri[-1][:]

        for i in range(m - 2, -1, -1):
            for j in range(i + 1):
                dp[j] = tri[i][j] + min(dp[j], dp[j + 1])

        return dp[0]

File: cpp/120_Triangle.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution {
public:
	int minimumTotal(vector<vector<int>>& tri) {
		int m = tri.size();

		// fill in as the last row of tri
		vector<int> dp = tri.back();

		// start at 2nd-last row now
		for (int i = m - 2; i >= 0; --i) {
			// stop at i trick since its a triangle
			for (int j = 0; j <= i; ++j) {
				// cost is trivally the cur index value
				// + best path looking down, i.e.
				// dp[i + 1][j] or dp[i + 1][j + 1] in a backwards manner
				dp[j] = tri[i][j] + min(dp[j], dp[j + 1]);
			}
		}

		return dp[0];
	}
};


